src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -729,12 +757,31 @@ int implGCMCrypt(ByteBuffer src, ByteBuffer dst) {
             int len;
 
             if (src.hasArray() && dst.hasArray()) {
-                ByteBuffer ct = (encryption ? dst : src);
-                len = GaloisCounterMode.implGCMCrypt(src.array(),
-                    src.arrayOffset() + src.position(), srcLen,
-                    ct.array(), ct.arrayOffset() + ct.position(),
-                    dst.array(), dst.arrayOffset() + dst.position(),
-                    gctr, ghash);
+                byte[] array;
+                if (encryption) {
+                    array = dst.array();
+                    int ofs = dst.arrayOffset() + dst.position();
+                    len = GaloisCounterMode.implGCMCrypt(src.array(),
+                        src.arrayOffset() + src.position(), srcLen,
+                        array, ofs, array, ofs, gctr, ghash);
+                } else {
+                    array = src.array();
+                    // Don't use 'inPlaceArray' here as the src maybe ibuffer
+                    // eventhough doFinal() may have been given in-place heap
+                    // bytebuffers
+                    if (array == dst.array()) {
+                        len = GaloisCounterMode.implGCMCrypt(array,
+                            src.arrayOffset() + src.position(), srcLen,
+                            null, 0, array, dst.arrayOffset() + dst.position(),
+                            gctr, ghash);
+                    } else {
+                        int ofs = src.arrayOffset() + src.position();
```

Comment:
The line 774 case only uses the calculated value once.  I'm avoiding the unnecessary store & load operations when it is set the value to a variable. I see it when I run  javap -c to view the bytecode.  It's purely an optimization for performance.   I do set it to the variable in the line 778 case because calculating the value twice is probably more expensive than once with the store & load operations.

Reply:
I see.